fsspec/spec.py

```diff
@@ -808,7 +818,7 @@ def rm(self, path, recursive=False, maxdepth=None):
     def _parent(cls, path):
         path = cls._strip_protocol(path.rstrip("/"))
         if "/" in path:
-            return cls.root_marker + path.rsplit("/", 1)[0]
+            return os.path.join(cls.root_marker, path.rsplit("/", 1)[0])
```

Comment:
> Also, I suspect that this would make "//path/..."

Exactly, that's why I fix the `_parent` method.

> If there was a specific error that you saw, it would be good to get that as a separate test.

In theory, it's correct. But
1. The `_parent` method is private, I think the private method is unstable, testing it is not a good idea;
2. In fact, the `_parent` method only causes errors in the `zip.py` and `ftp.py` files in which `root_marker` is not empty. It works well in `spec.py`, so you can't test the actual problem in `spec.py`, or it's not easy to write a straightforward test.
So I think it's ok to indirectly test the _parent method in ftp.py for now.

But in the long run, all the logic about directories included `_parent` should be refactored, may be extracted into a utils file, and tested well. For example, we have too much code like this:
```
            elif (
                path
                and len(path) < len(p.strip("/"))
                and all(
                    (a == b) for a, b in zip(path.split("/"), p.strip("/").split("/"))
                )
            ):
```

Reply:
Oh, and if we officially support windows and mac osx, we should add them to our CI environment. I may post a pull request for that later.

Reply:
That would be appreciated